File: dataStructures/hash_table.rb
class HashTable

  def initialize
    @size = 17
    @keys = [nil]*@size
    @data = [nil]*@size
  end

  def put(key, data)
    hash_value = hash(key)
    if (@keys[hash_value].nil?)
      # no collision, found spot, insert new data
      @keys[hash_value] = key
      @data[hash_value] = data
    elsif (@keys[hash_value] == key)
      # no collision, found spot, replace old data
      @data[hash_value] = data
    else
      probe = 1
      hash_value = rehash(hash_value, probe)
      while (!@keys[hash_value].nil? and @keys[hash_value] != 'deleted' and @keys[hash_value] != key)
        probe += 1
        hash_value = rehash(hash_value, probe)
      end
      if (@keys[hash_value].nil? or @keys[hash_value] == 'deleted')
        # found open slot
        @keys[hash_value] = key
        @data[hash_value] = data
      elsif (@keys[hash_value] == key)
        # found spot, replace old data
        @data[hash_value] = data
      end
    end
  end

  def get(key)
    hash_value = hash(key)
    if (@keys[hash_value] == key)
      # found slot, return data
      return @data[hash_value]
    else
      probe = 1
      new_hash = rehash(hash_value, probe)
      while (@keys[new_hash] != key and !@keys[new_hash].nil? and new_hash != hash_value)
        probe += 1
        new_hash = rehash(new_hash, probe)
      end

      if (@keys[new_hash] == key)
        # found slot, return data
        return @data[new_hash]
      else
        # one of other two conditions, data not present
        return nil
      end
    end
  end

  def delete(key)
    hash_value = hash(key)
    if (@keys[hash_value] == key)
      # found slot, mark as delete, set data to nil
      @keys[hash_value] = 'deleted'
      @data[hash_value] = nil
    else
      probe = 1
      new_hash = rehash(hash_value, probe)
      while (@keys[new_hash] != key and !@keys[new_hash].nil? and new_hash != hash_value)
        probe += 1
        new_hash = rehash(new_hash, probe)
      end

      if (@keys[new_hash] == key)
        # found slot, mark as deleted, set data to nil
        @keys[new_hash] = 'deleted'
        @data[new_hash] = nil
      else
        # trying to delete value that does not exist
        return nil
      end
    end
  end

  def [](key)
    get(key)
  end

  def []=(key, data)
    put(key, data)
  end
      

  private

  def hash(key)
    return key % @size
  end

  def rehash(old, probe)
    return (old + probe**2) % @size
    #return (old+1) % @size
  end


end


begin
  h = HashTable.new
  (0..16).each do |i|
    h[i] = 'k'
  end
  h.delete(15)
  puts h.inspect
  puts ""
  h[15] = 'HELLO'
  puts ""
  puts h.inspect
end

File: dataStructures/hashTable.py
class HashTable:

    def __init__(self, size):
        self.size = size
        self.keys = [None]*size
        self.data = [None]*size

    def put(self, key, data):
        hashValue = self._hash(key)
        if (self.keys[hashValue] == None):
            # no collision, found empty slot, so insert
            self.keys[hashValue] = key
            self.data[hashValue] = data
        elif (self.keys[hashValue] == key):
            # no collision, found spot, replace old data
            self.data[hashValue] = data
        else:
            hashValue = self._rehash(hashValue)
            while (self.keys[hashValue] != None and self.keys[hashValue] != key and self.keys[hashValue] != 'deleted'):
                hashValue = self._rehash(hashValue)
            if (self.keys[hashValue] == None or self.keys[hashValue] == 'deleted'):
                # found empty slot, insert data
                self.keys[hashValue] = key
                self.data[hashValue] = data
            else:
                # found slot, replace data
                self.data[hashValue] = data


    def get(self, key):
        hashValue = self._hash(key)
        found = False
        stop = False
        startPos = hashValue
        while (self.keys[hashValue] != None and not found and not stop):
            if (self.keys[hashValue] == key):
                found = True
            else:
                hashValue = self._rehash(hashValue)
                if (hashValue == startPos):
                    stop = True
        if (found):
            return self.data[hashValue]
        else:
            return None

    def delete(self, key):
        hashValue = self._hash(key)
        found = False
        stop = False
        startPos = hashValue
        while (self.keys[hashValue] != None and not found and not stop):
            if (self.keys[hashValue] == key):
                found = True
            else:
                hashValue = self._rehash(hashValue)
                if (hashValue == startPos):
                    stop = True
        if (found):
            self.keys[hashValue] = 'deleted'
            self.data[hashValue] = None
        else:
            return False

    def _hash(self, key):
        return key % self.size

    def _rehash(self, hashValue):
        return (hashValue+1) % self.size
            


h = HashTable(11)
h.put(1,3)
h.put(12,5)
h.put(23, 78)
print h.keys
h.delete(12)
print h.get(1)
print h.get(23)
print h.keys
h.put(34, 35)
print h.keys
print h.data
#h.put(5,6)
#h.put(7,9)

